packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1795,24 +1781,32 @@ class OpacityLayer extends ContainerLayer {
       return true;
     }());
 
-    if (enabled)
+    final int realizedAlpha = alpha!;
+    // The type assertions work because the [alpha] setter nulls out the
+    // engineLayer if it would have changed type (i.e. changed to or from 255).
```

Comment:
Inter-method dependencies for correctness are weaker than just having this method deal with the wrong layer type that it knows happens because it is the method that causes it.

Reply:
The asserts in here guard against this getting broken, and there is test coverage that would trigger them should it break. I did this becuase it's on a hot path.